### PR TITLE
Inclusion of LArQL model, a model for correlated charge and light, dependent on the electric field. More details can be found at https://drive.google.com/file/d/1zG5lLqMxACcvuJqQXhLnQ8Cnmz2YEZPU/view?usp=sharing

### PR DESCRIPTION
**Pull request by lpaulucc**

### CODE REVIEW

**Comment by knoepfel** [comment]
The new FHiCL parameters do not have default values.  Do you expect this to cause a breaking change?

**Comment by lpaulucc** [comment]
The new FHICL parameters are defined in
larsim/Simulation/simulationservices.fcl
<https://github.com/LArSoft/larsim/pull/66/files#diff-cdf2515e99f87521ce45274eef45db10f2b1c1e1b052782b62759fa3f6de5d14>

Em seg., 29 de mar. de 2021 às 12:41, Kyle Knoepfel <
***@***.***> escreveu:

**Comment by knoepfel** [comment]
Understood, @lpaulucc.  However, users must be sure to include that FHiCL file to ensure their workflows do not break.  Do enough people include that file that it's not necessary to mention the breaking change?

**Comment by lpaulucc** [comment]
Ah, ok, Kyle. Sorry for the confusion.
I made those changes in larsim and simulated events using dunetpc without
making any change to the code inside dunetpc. I had no problems since the
file simulationservices_dune.fcl includes simulationservices.fcl. The same
for simulationservices_sbnd.fcl.
What I see is
that larsim/larsim/Simulation/simulationservices_lartpcdetector.fcl

simulationservices_jp250L.fcl

simulationservices_bo.fcl

simulationservices_argoneut.fcl
all include larsim/Simulation/simulationservices.fcl. I am not sure about
other experiments.
I assume that if the simulation services for the specific experiment does
not include larsim/Simulation/simulationservices.fcl, then it would break
things. Do you think the parameters should also be included in
larcoreobj/SimpleTypesAndConstants/PhysicalConstants.h?


Em seg., 29 de mar. de 2021 às 13:37, Kyle Knoepfel <
***@***.***> escreveu:

**Comment by lgarren** [comment]
trigger build

**Comment by knoepfel** [comment]
@lpaulucc, let's not worry about including the parameters in `larcoreobj/SimpleTypesAndConstants/PhysicalConstants.h` yet.  As long as the CI tests return back okay, then it's probably fine to go ahead and just merge this in.

**Comment by lgarren** [comment]
uboone CI
https://dbweb8.fnal.gov:8443/LarCI/app/ns:uBooNE/view_builds/index?offset=0&builds=uboone_ci/6048&builds=

**Comment by lgarren** [comment]
CI results appear to be normal

File: larsim/IonizationScintillation/ISCalcCorrelated.h [hunk 1]
```diff
@@ -48,14 +50,23 @@ namespace larg4 {
...
+    bool fUseModLarqlRecomb;  ///< from LArG4Parameters service
 
     const spacecharge::SpaceCharge* fSCE;
     const detinfo::LArProperties* fLArProp;
 
     void CalcIon(sim::SimEnergyDeposit const& edep);
     void CalcScint(sim::SimEnergyDeposit const& edep);
     double GetScintYieldRatio(sim::SimEnergyDeposit const& edep);
+    double chi0(double dEdx);
+    double fcorr(double EF, double dEdx);
```
Comment (absolution1) [inline larsim/IonizationScintillation/ISCalcCorrelated.h]: Would it be possible to use more descriptive function names and make the capitalisation consistent with the rest of the codebase?

File: larsim/LegacyLArG4/ISCalculationCorrelated.h [hunk 2]
```diff
@@ -47,7 +49,17 @@ namespace larg4 {
...
+    double fLarqlChi0B;       ///< from LArG4Parameters service
+    double fLarqlChi0C;       ///< from LArG4Parameters service
+    double fLarqlChi0D;       ///< from LArG4Parameters service
+    double fLarqlAlpha;       ///< from LArG4Parameters service
+    double fLarqlBeta;        ///< from LArG4Parameters service
+    bool fUseModBoxRecomb;    ///< from LArG4Parameters service
+    bool fUseModLarqlRecomb;  ///< from LArG4Parameters service
+
+    double chi0(double dEdx);
+    double fcorr(double EF, double dEdx);
```
Comment (absolution1) [inline larsim/LegacyLArG4/ISCalculationCorrelated.h]: same comment as above about naming and capitalisation

File: larsim/Simulation/simulationservices.fcl [hunk 3]
```diff
@@ -42,7 +43,17 @@ standard_largeantparameters:
  ModBoxA: 0.930
  ModBoxB: 0.212
 
- # The following parameters specify details of wireplanes or similar
+ #* Recombination factor coefficients for LArQL can be found in https://drive.google.com/file/d/1zG5lLqMxACcvuJqQXhLnQ8Cnmz2YEZPU/view?usp=sharing
```
Comment (absolution1) [inline larsim/Simulation/simulationservices.fcl]: are the coefficients searchable elsewhere?  This comment relies on the googledrive link remaining valid...

File: larsim/IonizationScintillation/ISCalcCorrelated.h (hunk 1 above)
Comment (absolution1) [reply larsim/IonizationScintillation/ISCalcCorrelated.h]: I guess the other functions also pass const arguments as well.  If you update the naming/capitalisation, you could consider passing the arguments as const values too.

**Review by absolution1 — COMMENTED** [review]
I've left a couple of comments.  Most are stylistic and optional.
The last comment, regarding the googledrive link, should be addressed before merging if possible.

**Comment by absolution1** [comment]
BTW, Lancaster University is about to close for Easter so I'll be on mandated vacation time for the next week from today.
So, I am happy to stamp approval now but it would be good to at least get opinions on changing the google drive link

**Comment by absolution1** [comment]
approve

**Comment by lpaulucc** [comment]
Sorry for the delay, starting the day with many meetings already.

***@***.**** commented on this pull request.
>
> I've left a couple of comments. Most are stylistic and optional.
> The last comment, regarding the googledrive link, should be addressed
> before merging if possible.
> ------------------------------
>
> In larsim/IonizationScintillation/ISCalcCorrelated.h
> <https://github.com/LArSoft/larsim/pull/66#discussion_r604736549>:
>
> > +    double chi0(double dEdx);
> +    double fcorr(double EF, double dEdx);
>
> Would it be possible to use more descriptive function names and make the
> capitalisation consistent with the rest of the codebase?
>
Regarding the naming of the functions, those are the names currently
described in the model. I would need to check with the other collaborators
a possible change. I can update the capitalisation, no problem. I Will do
that today.



> ------------------------------
>
> In larsim/LegacyLArG4/ISCalculationCorrelated.h
> <https://github.com/LArSoft/larsim/pull/66#discussion_r604737000>:
>
> > +    double chi0(double dEdx);
> +    double fcorr(double EF, double dEdx);
>
> same comment as above about naming and capitalisation
> ------------------------------
>

 Same as above.

In larsim/Simulation/simulationservices.fcl
> <https://github.com/LArSoft/larsim/pull/66#discussion_r604737676>:
>
> > @@ -42,7 +43,17 @@ standard_largeantparameters:
>   ModBoxA: 0.930
>   ModBoxB: 0.212
>
> - # The following parameters specify details of wireplanes or similar
> + #* Recombination factor coefficients for LArQL can be found in https://drive.google.com/file/d/1zG5lLqMxACcvuJqQXhLnQ8Cnmz2YEZPU/view?usp=sharing
>
> are the coefficients searchable elsewhere? This comment relies on the
> googledrive link remaining valid...
>
I would suggest, if that is ok by everyone, that I update the wiki link
https://cdcvs.fnal.gov/redmine/projects/larsoft/wiki/Photon_and_Ionization_Electron_Simulation
including a brief discussion on the modified Birks and Box model and LArQL,
including the Separate and Correlated model for scintillation. I can also
include this document in docdb but I am not sure which one would be better
so that it could be accessible to everyone. Then the drive link could be
replaced by the docdb one. Would that be ok? Can anyone suggest an
appropriate docdb?


> In larsim/IonizationScintillation/ISCalcCorrelated.h
> <https://github.com/LArSoft/larsim/pull/66#discussion_r604738467>:
>
> > +    double chi0(double dEdx);
> +    double fcorr(double EF, double dEdx);
>
> I guess the other functions also pass const arguments as well. If you
> update the naming/capitalisation, you could consider passing the arguments
> as const values too.
>
I could not quite follow this one... Do you mean that it would be better to
be

double const dEdx = (ds <= 0.0) ? 0.0 : energy_deposit / ds;
double const EFieldStep = EFieldAtStep(detProp.Efield(), edep);

double chi0(double const dEdx)?

**Comment by lgarren** [comment]
@lpaulucc  Yes, please add  new  LArSoft wiki page from https://cdcvs.fnal.gov/redmine/projects/larsoft/wiki/Photon_and_Ionization_Electron_Simulation with the information now on the google drive link.

File: larsim/IonizationScintillation/ISCalcCorrelated.h (hunk 1 above)
Comment (knoepfel) [reply larsim/IonizationScintillation/ISCalcCorrelated.h]: The issue of `const` correctness is a long discussion.  The general gist: in a function _declaration_, there is no benefit in marking variables passed in by value as `const`.  However, there _is_ a benefit to marking such variables as `const` in  the function _definition_.  So Dom's comment about `const` variables is partially correct.  A good pattern to follow:

```.cpp
int foo(int bar); //  Function declaration - 'bar' is not marked const
int foo(int const bar) { return bar + 1; } // Function definition - 'bar' is marked const
```

**Comment by lgarren** [comment]
@lpaulucc LArSoft has no docdb, which is why I suggested adding a new wiki page with the information now in google drive.  You can attach a pdf document to the wiki if necessary, but this is not ideal.  
Also, about names:  Longer names can help avoid confusion and possible code conflicts if the same short name is used for different purposes in different contexts.

**Comment by lpaulucc** [comment]
Thanks everyone. I will make the needed corrections and push it later today.

Em qua., 31 de mar. de 2021 às 12:58, Lynn Garren ***@***.***>
escreveu:

**Comment by lgarren** [comment]
trigger build

**Comment by lgarren** [comment]
approve

**Pull request merged**